kernel/kernbench.py

```diff
@@ -0,0 +1,161 @@
+        process.system(build_string, allow_output_check='None', shell=True)
+        if not os.path.isfile('vmlinux'):
+            self.error("No vmlinux found, kernel build failed")
+
+    @staticmethod
+    def to_seconds(time_string):
+        """
+        Converts a string in M+:SS.SS format to S+.SS
+        """
+        elts = time_string.split(':')
+        if len(elts) == 1:
+            return time_string
+        return str(int(elts[0]) * 60 + float(elts[1]))
+
+    def extract_all_time_results(self, results_string):
+        """
+        Extract user, system, and elapsed times into a list of tuples
+        """
+        pattern = re.compile(r"(.*?)user (.*?)system (.*?)elapsed")
+        results = []
+        for result in pattern.findall(results_string):
+            results.append(tuple([self.to_seconds(elt) for elt in result]))
+        return results
+
+    def setUp(self):
+        """
+        Setting up the env for the kernel building
+        """
+        smg = SoftwareManager()
+        detected_distro = distro.detect()
+        deps = ['gcc', 'make', 'automake', 'autoconf', 'time']
+        if 'Ubuntu' in detected_distro.name:
+            deps.extend(['libpopt0', 'libc6', 'libc6-dev', 'libpopt-dev',
+                         'libcap-ng0', 'libcap-ng-dev', 'elfutils', 'libelf1',
+                         'libnuma-dev', 'libfuse-dev'])
+        elif 'SuSE' in detected_distro.name:
+            deps.extend(['libpopt0', 'glibc', 'glibc-devel',
+                         'popt-devel', 'libcap1', 'libcap-devel',
+                         'libcap-ng-devel'])
+        # FIXME: "redhat" as the distro name for RHEL is deprecated
+        # on Avocado versions >= 50.0.  This is a temporary compatibility
+        # enabler for older runners, but should be removed soon
+        elif detected_distro.name in ['centos', 'fedora', 'rhel', 'redhat']:
+            deps.extend(['popt', 'glibc', 'glibc-devel', 'libcap-ng',
+                         'libcap', 'libcap-devel', 'elfutils-libelf',
+                         'elfutils-libelf-devel', 'openssl-devel'])
+
+        for package in deps:
+            if not smg.check_installed(package) and not smg.install(package):
+                self.cancel(package + ' is needed for the test to be run')
+        self.kernel_version = platform.uname()[2]
+        self.iterations = self.params.get('runs', default=1)
+        self.threads = self.params.get(
+            'cpus', default=2 * cpu.online_cpus_count())
+        self.version = self.params.get('version', default='4.6')
+        self.config_path = '/boot/config-' + self.kernel_version
+
+    def test(self):
+        """
+        Kernel build Test
+        """
+        # Initialize kernel object
+        kern = kernel.KernelBuild(
+            self.version, config_path=self.config_path,
+            work_dir=None, data_dirs=None)
+        # Setting the default kernel URL
+        kern.URL = 'https://www.kernel.org/pub/linux/kernel/v4.x/'
```

Comment:
i think all from line no 128 to 134 can be wrap in such way as using fetch_asset , than extract in your kernel build path 

```
location = ["https://github.com/torvalds/linux/archive/master.zip"]
        tarball = self.fetch_asset("kselftest.zip", locations=location,
                                   expire='1d')
        archive.extract(tarball, self.srcdir)
        self.buldir = os.path.join(self.srcdir, 'linux-master')
        self.sourcedir = os.path.join(self.buldir, self.testdir)
```

as well url can be managed from yaml file

Reply:
sure sir